### PR TITLE
Remove reliance on discordcr, other fixes

### PR DESCRIPTION
**Pull request by voximity**
This PR includes the following:
* discordcr is removed as a dependency
* Cleans up and makes the code a bit more consistent
* Fixes timestamp comparisons in `reader.cr`
* Adds an embed display style for rendering post information
* Adds `DisplayStyle` enum to choose between `DisplayStyle::Text` and `DisplayStyle::Embed`. Configure in `config.json`'s `display_style` option.

### CODE REVIEW

File: src/reader.cr [hunk 1]
```diff
@@ -1,34 +1,38 @@
...
+      current_date = "#{@post_path}/#{Time.now.to_s "%F"}"
+      File.write(@post_record_path, current_date)
+      current_date
     end
 
     private def write_post_record(string : String)
       File.write(@post_record_path, string)
     end
 
+    private def timestamp_of_post(string : String)
```
Comment (z64) [inline src/reader.cr]: It is not necessary to extract the date from the post. We don't care about when the article was posted, only that we have or haven't posted it.

The articles will *always* be in lexical order because they start with an ISO 8601 date. `Dir.glob(pattern).sort` ensures we walk through them in order. Our last known filename is then just compared with `#>`.

File: .gitignore [hunk 2]
```diff
@@ -3,3 +3,4 @@
 /bin/
 /.shards/
 *.dwarf
+last_post.txt
```
Comment (z64) [inline .gitignore]: This is not necessary

File: src/config.cr [hunk 3]
```diff
@@ -1,4 +1,9 @@
 module CrystalNews
+  enum DisplayStyle
```
Comment (z64) [inline src/config.cr]: I don't think the display style should be configurable. Not worth the complexity for something so small.

File: src/crystal_news.cr [hunk 4]
```diff
@@ -27,6 +26,10 @@ module CrystalNews
     end
   end
 
+  def author_url(name)
```
Comment (z64) [inline src/crystal_news.cr]: Can this be moved to `BlogPost`?

And preferably `def author_avatar`

File: src/discord.cr [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+      @image_url = nil,
+      @thumbnail_url = nil,
+      @author = nil
+    ); end
+  end
+
+  struct DiscordMessage
+    include JSON::Serializable
+
+    @[JSON::Field]
```
Comment (z64) [inline src/discord.cr]: These can be omitted. It's the opposite with `JSON::Serializable`; ivars must be explicitly ignored.

File: src/discord.cr [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+      )
+
+      initialize(
+        username: "Crystal Blog",
+        avatar_url: CRYSTAL_ICON,
+        embeds: [embed]
+      )
+    end
+
+    def initialize(blog_post : BlogPost, style : DisplayStyle)
```
Comment (z64) [inline src/discord.cr]: I believe multiple `initialize` is a bad pattern. There should be a single `initialize` method that directly sets all of that types ivars, and specialized `def self.new(other_type : T)` constructors that ultimately call the default constructor.

Multiple initialize methods complicates types and initialization overall, and makes things like creating instances of the struct for testing much more difficult.

File: src/run.cr [hunk 7]
```diff
@@ -1,14 +1,17 @@
 require "./crystal_news"
+include CrystalNews
```
Comment (z64) [inline src/run.cr]: Please don't pollute the global namespace! There is no problem with typing out the full path in this file.

File: src/webhook.cr [hunk 8]
```diff
@@ -1,14 +1,12 @@
+require "http/client"
+require "uri"
```
Comment (z64) [inline src/webhook.cr]: This require is not needed

File: src/webhook.cr [hunk 9]
```diff
@@ -21,16 +19,26 @@ module CrystalNews
...
     end
 
     def post(message : DiscordMessage)
-      @discord.execute_webhook(@id, @token,
-        username: message.username,
-        avatar_url: message.avatar_url,
-        content: message.content,
-        embeds: message.embeds)
+      response = HTTP::Client.post(
+        url: "https://discordapp.com/api/webhooks/#{@id}/#{@token}",
```
Comment (z64) [inline src/webhook.cr]: Please specify `v6`

File: src/webhook.cr [hunk 10]
```diff
@@ -21,16 +19,26 @@ module CrystalNews
...
+        headers: HTTP::Headers {"Content-Type" => "application/json"},
+        body: message.to_json
+      )
```
Comment (z64) [inline src/webhook.cr]: Somewhere, `response.success?` must be checked to log any errors that occured. `HTTP::Client` does not raise exceptions for client or server errors - it only raises socket-related exceptions where the request couldn't be completed. 

It seems you removed logging of the requests altogether - please restore them. If failures occur, they will be impossible to track down and understand. You can create an exception class that holds the `HTTP::Response` if you want, but either way this needs to happen.

File: src/webhook.cr [hunk 11]
```diff
@@ -1,14 +1,12 @@
...
-      client = Discord::Client.new("", logger: logger)
-      webhooks = webhooks.map do |config|
-        Webhook.new(client, config.id, config.token)
-      end
-      WebhookManager.new(webhooks, logger)
-    end
+    @webhooks : Array(Webhook)
 
-    def initialize(@webhooks : Array(Webhook), @logger : Logger)
+    def initialize(webhooks : Array(WebhookConfig), @logger : Logger)
```
Comment (z64) [inline src/webhook.cr]: Please revert the constructor change per my other comment

File: src/discord.cr [hunk 12]
```diff
@@ -0,0 +1,144 @@
+require "json"
+
+module CrystalNews
+  struct DiscordEmbedField
+    include JSON::Serializable
+    
+    property name : String
```
Comment (z64) [inline src/discord.cr]: None of these structs should be mutable.

File: src/reader.cr (hunk 1 above)
Comment (voximity) [reply src/reader.cr]: I was experiencing a lot of issues getting it to work right, so I implemented this. Besides, later in the code, this evaluation is necessary, to get dates to appear correctly on embeds. This functionality should be kept.

File: .gitignore (hunk 2 above)
Comment (voximity) [reply .gitignore]: Only added it for my own sake because I forgot to run it in the correct directory. Will fix.

File: src/config.cr (hunk 3 above)
Comment (voximity) [reply src/config.cr]: The idea was that you could add additional display styles and configure between them. It's already there and it's not bloating anything.

File: src/crystal_news.cr (hunk 4 above)
Comment (voximity) [reply src/crystal_news.cr]: Will fix.

File: src/discord.cr (hunk 5 above)
Comment (voximity) [reply src/discord.cr]: Will fix.

File: src/discord.cr (hunk 6 above)
Comment (voximity) [reply src/discord.cr]: I don't understand your reasoning here.

File: src/run.cr (hunk 7 above)
Comment (voximity) [reply src/run.cr]: This is an application, not a library. There is no harm in including the content of the application itself into the entry-level file.

File: src/webhook.cr (hunk 8 above)
Comment (voximity) [reply src/webhook.cr]: Will fix. Was confused by some docs.

File: src/webhook.cr (hunk 9 above)
Comment (voximity) [reply src/webhook.cr]: As ridiculously unlikely it is that this API path will change, I will specify it anyway. As soon as v6 becomes obsolete, the code will have to be manually changed. Will fix anyway.

File: src/webhook.cr (hunk 11 above)
Comment (voximity) [reply src/webhook.cr]: I will allow you to make this change yourself. I am against your reasoning.

File: src/webhook.cr (hunk 10 above)
Comment (voximity) [reply src/webhook.cr]: It is unnecessary to have error logging that in-depth. I will simply raise an exception with the response's body if it does not succeed.

File: src/discord.cr (hunk 12 above)
Comment (voximity) [reply src/discord.cr]: Will fix.

File: src/discord.cr (hunk 6 above)
Comment (z64) [reply src/discord.cr]: It's perspective from having written a lot of Crystal, and a lot of libraries.

The problem is that `initialize` in Crystal is *special*. The compiler "reserves" any method named `initialize` and says "this method must set up all instance variables". It also magically defines a `self.new` method with the same arguments - that "hidden" method is actually what does the allocation for the object, etc.

Now, sometimes you want an object to have multiple constructors. Maybe they take fewer arguments - say, unpacking an instance of an object from an encoded string.

This is a drawn out example, but hopefully illustrates the point a little:

```cr
struct Foo
  def initialize(string : String)
    int_strings = string.split(',')
    initialize(int_strings)
  end
  
  def initialize(int_strings : Array(String))
    ints = int_strings.map(&.to_i)
    initialize(ints)
  end
  
  def initialize(@ints : Array(Int32))
  end
end
```

Writing these constructors is a delicate dance because they are under scrutiny of the compiler's law of `initialize`., and often spreads out logic across multiple methods, making it hard to follow. Consider that you might have other constructors that do things like make network requests (who knows) - now that type's instance methods (post-construction) has become more difficult to test with networking logic intertwined.

What has been the best solution in complicated code bases is to define our own class-level constructors as `def self.new(other : OtherType)`, that wholly contains the conversion logic and avoids the dance between methods, and avoids some really bad errors with the compilers `initialize` rules: If you have several `initialize`, and you get the compiler's error that an ivar wasn't set up, you now have to jump through multiple hoops to find where an ivar wasn't set.

Ex:
```cr
struct Bar
  # Make a new bar from a String
  def self.new(string : String)
    ints = string.split(',').map(&.to_i)
    Bar.new(ints)
  end

  # Make a new bar from a BAz
  def self.new(baz : Baz)
    # somehow, get an array of ints out of Baz
    Bar.new(ints)
  end

  def initialize(@ints : Array(Int32))
  end
end
```

Additionally, when we want to test this type, we can just initialize it right from its default constructor and skip any intermediate construction:
```cr
Bar.new([1, 2, 3])
```

Overall this is a much easier rule to apply consistently and scale instead of having multiple layers of method calls on a whim in order to set up a type.

File: src/webhook.cr (hunk 10 above)
Comment (z64) [reply src/webhook.cr]: Being able to audit the application is respectable and important to people who host it. Since the application supports multiple webhooks, say one of the webhooks gets deleted - if we just raise an exception, this crashes the whole distribution. Instead we log that the webhook 404'd so it can be pruned in auditing later, and other webhooks in the configuration can still succeed.

File: src/discord.cr (hunk 6 above)
Comment (voximity) [reply src/discord.cr]: I have never had a problem overloading `initialize` with varying parameter types, but I do appreciate this explanation. Intermediate construction can be skipped implicitly if you overload `initialize` as there is no newly defined `new` method that does this for you. This is merely preference, and I don't see anything wrong with my method. However, I will make the change to keep the code consistent with yours.

Your first example certainly does not have to result in calling the same `initialize` method each time. I would only do that if the final method had some logic I was not willing to include in other overloaded methods to save space. In that regard, I would initialize the object just as the last method is doing. Thus,

```cr
struct Foo
  def initialize(string : String)
    @ints = string.split ','
  end
  
  def initialize(int_strings : Array(String))
    @ints = int_strings.map(&.to_i)
  end
  
  def initialize(@ints : Array(Int32))
  end
end
```

has no problem because there isn't a central `initialize` method each one refers to.

Constructors shouldn't make network events, or anything that might delay or hang the current Fiber. In this circumstance, this functionality *should* be moved to a class-level method constructor. `initialize` constructors should be responsible for mutating input parameters and setting properties of the object in question, but should not commit any actions aside from that. Again, that functionality should be saved for either a class-level method or a separate instance-level method.

File: src/webhook.cr (hunk 11 above)
Comment (voximity) [reply src/webhook.cr]: Fixed as per our discussion in a prior comment.

File: src/discord.cr (hunk 6 above)
Comment (z64) [reply src/discord.cr]: > Your first example certainly does not have to result in calling the same initialize method each time.

>  Constructors shouldn't make network events, or anything that might delay or hang the current Fiber. 

I know my example wasn't exactly the same as what we're discussing here, but my point was that "initializing a type from other types should be done outside of initialize in `new` methods" is an easier rule to apply, rely on, and scale (keep consistent) than to bare all of those considerations in mind. Hell, they don't even have to be called `new`, `def self.from(other : T)`, etc..

*If* someone did write some kind of wild constructor that had IO requirements, or otherwise complicated `initialize`, it wouldn't affect things like test cases for that struct that can always build the type from primitives.

Aside: Crystal, unfortunately (and necessarily so), doesn't have things like struct instance literals like C, Rust, etc. that doesn't require stepping through a function to create an instance.. It's really handy when testing more complicated types in bigger systems.

Crystal doesn't enforce a way to do this, so we (users) are forced to make decisions and agreements on how to do things like this.. :disappointed: 

Anyways; thanks for your consideration.
